apps/language_server/lib/language_server/providers/code_action/helpers.ex

```diff
@@ -0,0 +1,53 @@
+defmodule ElixirLS.LanguageServer.Providers.CodeAction.Helpers do
+  alias ElixirLS.LanguageServer.Protocol.TextEdit
+  alias ElixirLS.LanguageServer.Providers.CodeMod.Ast
+  alias ElixirLS.LanguageServer.Providers.CodeMod.Text
+
+  @spec update_line(TextEdit.t(), non_neg_integer()) :: TextEdit.t()
+  def update_line(
+        %TextEdit{range: %{"start" => start_line, "end" => end_line}} = text_edit,
+        line_number
+      ) do
+    %TextEdit{
+      text_edit
+      | range: %{
+          "start" => %{start_line | "line" => line_number},
+          "end" => %{end_line | "line" => line_number}
+        }
+    }
+  end
+
+  @spec to_one_line_string(Ast.t()) :: {:ok, String.t()} | :error
+  def to_one_line_string(updated_ast) do
+    updated_ast
+    |> Ast.to_string()
+    # We're dealing with a single error on a single line.
+    # If the line doesn't compile (like it has a do with no end), ElixirSense
+    # adds additional lines to documents with errors. Also, in case of a one-line do,
+    # ElixirSense creates do with end from the AST.
+    |> maybe_recover_one_line_do(updated_ast)
```

Comment:
If the original line has one-line do, after conversion to AST and conversion back to string, the output line has `do end`. Here I made a workaround that preserves one-line do. It is possible because the ASTs of the cases differ - just Macro.to_string/1 creates the same line out of both.